### PR TITLE
[Docker Check] add failed count memory metric

### PR DESCRIPTION
**Pull request by ziquanmiao**
### What does this PR do?
Extends Docker check to provide Failed Memory Count Metric

Containers have runtime constraints on resources like memory.
Failed Count is incremented every time memory allocation fails -- when a container's pre-set limit is hit

### Motivation

Customer request
Metric that is important for platform/app teams as an indicator for containers needing more memory

### Additional Notes

Anything else we should know when reviewing?

### CODE REVIEW

File: pkg/collector/corechecks/containers/docker.go [hunk 1]
```diff
@@ -201,6 +201,10 @@ func (d *DockerCheck) Run() error {
 				sender.Gauge("docker.mem.swap", float64(c.Memory.Swap), "", tags)
 			}
 
+			if c.Memory.MemFailCntPresent == true {
```
Comment (ziquanmiao) [inline pkg/collector/corechecks/containers/docker.go]: This part should be extending the docker check file to get the Container object's memfailed count metric

File: pkg/util/containers/metrics/cgroup_metrics.go [hunk 2]
```diff
@@ -104,6 +104,9 @@ func (c ContainerCgroup) Mem() (*CgroupMemStat, error) {
 			ret.TotalActiveFile = v
 		case "total_unevictable":
 			ret.TotalUnevictable = v
+		case "failcnt":
```
Comment (ziquanmiao) [inline pkg/util/containers/metrics/cgroup_metrics.go]: I think this is where the agent parses the statsfile and if it sees a value for failcnt will report it

Do we want to be reporting 0s

Comment (CharlyF) [reply pkg/util/containers/metrics/cgroup_metrics.go]: IMO if the metric is not available in the statsfile, we should not report it at all

File: pkg/util/containers/metrics/cgroup_metrics.go [hunk 3]
```diff
@@ -131,6 +131,18 @@ func (c ContainerCgroup) MemLimit() (uint64, error) {
 	return v, nil
 }
 
+// FailedMemoryCount returns the number of times this cgroup reached its memory limit, if it exists.
+// If the file does not exist or there is no limit, then this will default to 0
+func (c ContainerCgroup) FailedMemoryCount() (unit64, error){
+	v, err := c.ParseSingleStat("memory", "memory.failcnt")
+	if os.IsNotExist(err) {
```
Comment (ziquanmiao) [inline pkg/util/containers/metrics/cgroup_metrics.go]: Hey @CharlyF, a lot of the functions here submit 0 if does not exist like softmemlimit() and CPUNrThrottled

I think if we accept this as practice, i'll follow along for FailedMemoryCount, but I think we should update documentation to reflect 0 if null like docker.mem.soft_limit in https://docs.datadoghq.com/integrations/docker_daemon/#pagetitle

File: pkg/collector/corechecks/containers/docker.go [hunk 4]
```diff
@@ -200,7 +200,7 @@ func (d *DockerCheck) Run() error {
 			if c.Memory.SwapPresent == true {
 				sender.Gauge("docker.mem.swap", float64(c.Memory.Swap), "", tags)
 			}
-
+
```
Comment (hkaj) [inline pkg/collector/corechecks/containers/docker.go]: Let's remove these whitespaces. Also pro tip: you can configure your editor to do that automatically on file save :)

File: pkg/util/containers/container_linux.go [hunk 5]
```diff
@@ -40,7 +40,10 @@ func (c *Container) FillCgroupLimits() error {
 	if err != nil {
 		return fmt.Errorf("soft mem limit: %s", err)
 	}
-
+	c.MemFailCnt, err = c.cgroup.FailedMemoryCount()
+	if err != nill {
```
Comment (CharlyF) [inline pkg/util/containers/container_linux.go]: s/nill/nil

File: pkg/util/containers/metrics/cgroup_metrics.go [hunk 6]
```diff
@@ -131,6 +131,18 @@ func (c ContainerCgroup) MemLimit() (uint64, error) {
...
+		log.Debugf("Missing cgroup file: %s",
+			c.cgroupFilePath("memory", "memory.failcnt"))
+		return 0, nil
+	} else if err != nil {
+		return 0, err
+	}
```
Comment (CharlyF) [inline pkg/util/containers/metrics/cgroup_metrics.go]: you need to return `return v, nil` at the end

**Review by CharlyF — CHANGES_REQUESTED** [review]

**Review by CharlyF — APPROVED** [review]

**Pull request merged**